here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/models/storage/Result.java

```diff
@@ -58,11 +58,116 @@ public abstract class Result implements Typed, AutoCloseable {
   public <FEATURE, CODEC extends FeatureCodec<FEATURE, CODEC>> @NotNull ForwardCursor<FEATURE, CODEC> cursor(
       @NotNull FeatureCodecFactory<FEATURE, CODEC> codecFactory) throws NoCursor {
     if (cursor != null) {
+      if (cursor instanceof HeapCacheCursor) {
+        cursor = ((HeapCacheCursor<?, ?>) cursor).getOriginalCursor();
```

Comment:
I like this, but I think with this implementation you will override the cursor member, which is likely not what you want. Another thing is, when the session is committed, it might replace all pending cursors in all pending results with a heap-cached one and might need to remove the reference to the original cursor. So, I would say, we need to expect that getOriginalCursor returns null. In a nutshell, I would expect something like this as code:
```
var cursor = this.cursor;
if (cursor instanceof HeapCacheCursor) {
  cursor = ((HeapCacheCursor<?, ?>) cursor).getOriginalCursor();
  if (cursor == null) {
    cursor = this.cursor;
  }
```
This should prevent overriding the member, so keep the heap-cached-cursor and allow to use the original, if being available. Actually, I would assume that getOriginalCursor() does return null, when the cursor is closed (just currently a isClosed() method is missing, we need first to add it, I guess.

Reply:
One question that came into my mind is the following: When we created the heap-cached-cursor, isn't the actual forward only cursor already moved forward, which is very likely not what the user expects, when calling `cursor(factory)`, in other words, should we at all do the check?
I like, that you can get the original cursor and maybe we can offer a method to like `loadMore(limit)`to the mutable cursor, so that you can optionally add more into the heap-cache as long as possible.

Reply:
ForwardCursor position is before first, the psql rs position might be different because hasNext() method has to load something from rs. Maybe Id don't see something but it think it's covered by tests.

Reply:
I have added return null in getOriginalCursor